core/consensus/babe/impl/babe_lottery_impl.cpp

```diff
@@ -15,99 +18,114 @@ namespace kagome::consensus::babe {
-      std::shared_ptr<BabeConfigRepository> babe_config_repo,
       std::shared_ptr<crypto::Hasher> hasher)
-      : vrf_provider_{std::move(vrf_provider)},
-        hasher_{std::move(hasher)},
-        logger_{log::createLogger("BabeLottery", "babe_lottery")} {
+      : logger_(log::createLogger("BabeLottery", "babe_lottery")),
+        config_repo_(std::move(config_repo)),
+        session_keys_(std::move(session_keys)),
+        vrf_provider_(std::move(vrf_provider)),
+        hasher_(std::move(hasher)) {
+    BOOST_ASSERT(config_repo_);
+    BOOST_ASSERT(session_keys_);
     BOOST_ASSERT(vrf_provider_);
     BOOST_ASSERT(hasher_);
     BOOST_ASSERT(logger_);
-    BOOST_ASSERT(babe_config_repo);
     epoch_ = std::numeric_limits<EpochNumber>::max();
   }
 
-  void BabeLotteryImpl::changeEpoch(EpochNumber epoch,
-                                    const Randomness &randomness,
-                                    const Threshold &threshold,
-                                    const crypto::Sr25519Keypair &keypair) {
-    SL_TRACE(logger_,
-             "Epoch changed "
-             "FROM epoch {} with randomness {} TO epoch {} with randomness {}",
-             epoch_,
-             randomness_,
-             epoch,
-             randomness);
-
+  bool BabeLotteryImpl::changeEpoch(EpochNumber epoch,
+                                    const primitives::BlockInfo &best_block) {
     epoch_ = epoch;
-    randomness_ = randomness;
-    threshold_ = threshold;
-    keypair_ = keypair;
-  }
 
-  EpochNumber BabeLotteryImpl::getEpoch() const {
-    return epoch_;
-  }
+    auto config_res = config_repo_->config(best_block, epoch);
+    [[unlikely]] if (config_res.has_error()) {
+      SL_ERROR(logger_,
+               "Can not get epoch: {}; Epoch was not changed",
+               config_res.error());
+      return false;
+    }
+    auto &config = *config_res.value();
 
-  std::optional<crypto::VRFOutput> BabeLotteryImpl::getSlotLeadership(
-      SlotNumber slot) const {
-    BOOST_ASSERT_MSG(epoch_ != std::numeric_limits<uint64_t>::max(),
-                     "Epoch must be initialized before this point");
+    keypair_ = session_keys_->getBabeKeyPair(config.authorities);
 
-    primitives::Transcript transcript;
-    prepareTranscript(transcript, randomness_, slot, epoch_);
+    randomness_ = config.randomness;
+
+    auth_number_ = config.authorities.size();
 
-    auto res = vrf_provider_->signTranscript(transcript, keypair_, threshold_);
+    threshold_ = babe::calculateThreshold(
+        config.leadership_rate, config.authorities, keypair_->second);
 
-    SL_TRACE(
-        logger_,
-        "prepareTranscript (leadership): randomness {}, slot {}, epoch {}{}",
-        randomness_,
-        slot,
-        epoch_,
-        res.has_value() ? " - SLOT LEADER" : "");
+    allowed_slots_ = config.allowed_slots;
 
-    return res;
+    SL_TRACE(logger_, "Epoch changed to epoch {}", epoch_);
+    return keypair_.has_value();
```

Comment:
If this function returns true when epoch has changed, does it makes sense to return true when keypair has value? What if we are running as syncing node?

Reply:
if we have no one key, it is one case of non-having key actual for epoch. In all such case we can not be validator.